### PR TITLE
Adapters to pass serializers as class members

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

File: zipline/src/commonMain/kotlin/app/cash/zipline/ZiplineApiMismatchException.kt [hunk 1]
```diff
@@ -1,3 +1,18 @@
+/*
+ * Copyright (C) 2022 Block, Inc.
```
Comment (oldergod) [inline zipline/src/commonMain/kotlin/app/cash/zipline/ZiplineApiMismatchException.kt:2]: I didn't see any announcement on this, we're good to switch? should we? is it optional?

**Review by oldergod — APPROVED** [review]
nice doc

File: zipline/src/commonMain/kotlin/app/cash/zipline/serializers.kt [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+ * ```kotlin
+ * val mySerializersModule = SerializersModule {
+ *   contextual(EchoService::class, ziplineServiceSerializer())
+ * }
+ * ```
+ *
+ * Next, use that `SerializersModule` when you create your Zipline instance, in both the host
+ * application and the Kotlin/JS code:
+ *
+ * ```
```
Comment (JakeWharton) [inline zipline/src/commonMain/kotlin/app/cash/zipline/serializers.kt]:
```suggestion
 * ```kotlin
```

File: zipline/src/commonMain/kotlin/app/cash/zipline/serializers.kt [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ * val zipline = Zipline.create(dispatcher, mySerializersModule)
+ * ```
+ *
+ * Finally, annotate the service property with `@Contextual`. This instructs Kotlin serialization to
+ * use the registered serializer.
+ *
+ * ```
```
Comment (JakeWharton) [inline zipline/src/commonMain/kotlin/app/cash/zipline/serializers.kt]:
```suggestion
 * ```kotlin
```

**Review by JakeWharton — APPROVED** [review]

File: zipline/src/engineTest/kotlin/app/cash/zipline/ZiplineServiceTest.kt [hunk 4]
```diff
@@ -115,4 +148,20 @@ internal class ZiplineServiceTest {
...
+
+  interface ServiceAndHashFactory : ZiplineService {
+    fun create(greeting: String): GreetingAndEchoService
+  }
+
+  class RealServiceAndHashFactory : ServiceAndHashFactory {
+    override fun create(greeting: String): GreetingAndEchoService {
+      return GreetingAndEchoService(greeting, GreetingService(greeting))
+    }
+  }
```
Comment (veyndan) [inline zipline/src/engineTest/kotlin/app/cash/zipline/ZiplineServiceTest.kt]:
```suggestion
  interface GreetingAndEchoServiceFactory : ZiplineService {
    fun create(greeting: String): GreetingAndEchoService
  }

  class RealGreetingAndEchoServiceFactory : GreetingAndEchoServiceFactory {
    override fun create(greeting: String): GreetingAndEchoService {
      return GreetingAndEchoService(greeting, GreetingService(greeting))
    }
  }
```

File: zipline/src/commonMain/kotlin/app/cash/zipline/serializers.kt [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+import app.cash.zipline.internal.bridge.ZiplineServiceAdapter
+import kotlinx.serialization.KSerializer
+
+/**
+ * Returns a [KSerializer] for [T] that performs pass-by-reference instead of pass-by-value. This is
+ * only necessary when a service is passed as a member of another serializable type; Zipline
+ * automatically does pass-by-reference for service parameters and return values.
+ *
+ * To use this, first register the serializer for your service in a `SerializersModule`.
```
Comment (veyndan) [inline zipline/src/commonMain/kotlin/app/cash/zipline/serializers.kt]:
```suggestion
 * To use this, first register the serializer for your service in a [kotlinx.serialization.modules.SerializersModule].
```

File: zipline/src/commonMain/kotlin/app/cash/zipline/serializers.kt [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+ *
+ * ```kotlin
+ * val mySerializersModule = SerializersModule {
+ *   contextual(EchoService::class, ziplineServiceSerializer())
+ * }
+ * ```
+ *
+ * Next, use that `SerializersModule` when you create your Zipline instance, in both the host
```
Comment (veyndan) [inline zipline/src/commonMain/kotlin/app/cash/zipline/serializers.kt]:
```suggestion
 * Next, use that [kotlinx.serialization.modules.SerializersModule] when you create your Zipline instance, in both the host
```

File: zipline/src/commonMain/kotlin/app/cash/zipline/serializers.kt [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+ * application and the Kotlin/JS code:
+ *
+ * ```
+ * val zipline = Zipline.create(dispatcher, mySerializersModule)
+ * ```
+ *
+ * Finally, annotate the service property with `@Contextual`. This instructs Kotlin serialization to
```
Comment (veyndan) [inline zipline/src/commonMain/kotlin/app/cash/zipline/serializers.kt]:
```suggestion
 * Finally, annotate the service property with [kotlinx.serialization.Contextual]. This instructs Kotlin serialization to
```

File: zipline/src/engineTest/kotlin/app/cash/zipline/ZiplineServiceTest.kt [hunk 8]
```diff
@@ -80,6 +84,35 @@ internal class ZiplineServiceTest {
...
+  @Test
+  fun registerSerializerToSerializeServicesAsMembers() = runBlocking {
+    val serializersModule = SerializersModule {
+      contextual(EchoService::class, ziplineServiceSerializer())
+    }
+    val (endpointA, endpointB) = newEndpointPair(this, serializersModule)
+
+    val serializer = serializersModule.serializer<EchoService>()
+    println(serializer)
+
```
Comment (veyndan) [inline zipline/src/engineTest/kotlin/app/cash/zipline/ZiplineServiceTest.kt]:
```suggestion
```

**Review by veyndan — APPROVED** [review]

File: zipline/src/commonMain/kotlin/app/cash/zipline/ZiplineApiMismatchException.kt (hunk 1 above)
Comment (swankjesse) [reply zipline/src/commonMain/kotlin/app/cash/zipline/ZiplineApiMismatchException.kt:2]: Yes. Use `Block, Inc.` for new files. Don‘t retroactively fix anything existing.

File: zipline/src/engineTest/kotlin/app/cash/zipline/ZiplineServiceTest.kt (hunk 4 above)
Comment (swankjesse) [reply zipline/src/engineTest/kotlin/app/cash/zipline/ZiplineServiceTest.kt]: Applied manually.

**Pull request merged**